translib/db/db_cursor.go

```diff
@@ -0,0 +1,343 @@
+//  pagination. [ Variable number of keys returned for every call to
+//  the ScanCursor API is ok. If fixed is needed, it may be implemented later.]
+
+// Duplicate Suppression? Redis can return duplicates. To supress them
+//  we need a cache. This will negate the Space complexity advantage of
+//  pagination. [ Duplicates to be suppressed. Currently the OnChange is
+//  getting all the keys first, so there is effectively a store of all the
+//  keys at some point. This will now be moved to a map/dictionary inside the
+//  ScanCursor to avoid duplicates.]
+
+////////////////////////////////////////////////////////////////////////////////
+//  Exported Types                                                            //
+////////////////////////////////////////////////////////////////////////////////
+
+// ScanCursor iterates over a set of keys
+
+type ScanCursor struct {
+	ts           *TableSpec
+	cursor       uint64
+	pattern      Key
+	count        int64
+	scanComplete bool
+	seenKeys     map[string]bool
+	lookAhead    []string // (TBD) For exactly CountHint # of keys
+	db           *DB
+	scnr         scanner
+}
+
+// ScanType type indicates the type of scan (Eg: KeyScanType, FieldScanType).
+type ScanType int
+
+const (
+	KeyScanType ScanType = iota // 0
+	FieldScanType
+)
+
+type ScanCursorOpts struct {
+	CountHint       int64  // Hint of redis work required
+	ReturnFixed     bool   // (TBD) Return exactly CountHint # of keys
+	AllowDuplicates bool   // Do not suppress redis duplicate keys
+	ScanType               // To mention the type of scan; default is KeyScanType
+	FldScanPatt     string // Field pattern to scan
+	AllowWritable   bool   // Allow on write enabled DB object; ignores tx cache
+}
+
+type scanner interface {
+	scan(sc *ScanCursor, countHint int64) ([]string, uint64, error)
+}
+
+type keyScanner struct {
+}
+
+type fieldScanner struct {
+	fldNamePattern string // pattern to match field name
+}
+
+func (scnr *keyScanner) scan(sc *ScanCursor, countHint int64) ([]string, uint64, error) {
+	return sc.db.client.Scan(sc.cursor,
+		sc.db.key2redis(sc.ts, sc.pattern), countHint).Result()
+}
+
+func (scnr *fieldScanner) scan(sc *ScanCursor, countHint int64) ([]string, uint64, error) {
+	key := sc.ts.Name
+	if len(sc.pattern.Comp) > 0 {
+		key = sc.db.key2redis(sc.ts, sc.pattern)
+	}
+	return sc.db.client.HScan(key, sc.cursor, scnr.fldNamePattern, countHint).Result()
+}
+
+////////////////////////////////////////////////////////////////////////////////
+//  Exported Functions                                                        //
+////////////////////////////////////////////////////////////////////////////////
+
+// NewScanCursor Factory method to create ScanCursor; Scan cursor will not be supported only for write enabled DB
+func (d *DB) NewScanCursor(ts *TableSpec, pattern Key, scOpts *ScanCursorOpts) (*ScanCursor, error) {
+	if glog.V(3) {
+		glog.Info("NewScanCursor: Begin: ts: ", ts, " pattern: ", pattern,
```

Comment:
Please refer -> https://pkg.go.dev/github.com/golang/glog#V

1. if glog.V(2) { glog.Info("log this") }
2. glog.V(2).Info("log this")

The second form is shorter but **the first is cheaper** if logging is off because it does not evaluate its arguments.